### PR TITLE
Standalone build of c3d as a separate executable 

### PR DESCRIPTION
**Pull request by msprotz**
I was able to fix the linking issues and work out the build. This worked on OSX and Linux. Please note that:

    if you have an existing "out" directory, you need to blast it
    from quackyducky/c3d, you now need to do: mkdir out && cd out && cmake .. -DLLVM_BUILD_ROOT=... && make -j

@tahina-pro does this change anything regarding the build?

Thanks!

### CODE REVIEW

**Comment by msprotz** [comment]
Pushed further fixes, this also fixes the build on linux and removes the OSX-specific linking procedure.

**Comment by tahina-pro** [comment]
I just fixed the build (`target` was missing from the Docker image.)
I also made `./build.sh` run the standalone test from outside of the Docker image, to be sure that `clang-c3d` can run without clang.

It now remains to make the test work.

**Comment by msprotz** [comment]
Wonderful, much appreciated Tahina. I'll merge as soon as the test is green.

When I push to this PR are the c3d tests run?

**Comment by tahina-pro** [comment]
The only test run is `./clang-c3d tests/basic0.h`, at the end of `c3d/build.sh`
Please feel free to add more there if you want.

**Comment by tahina-pro** [comment]
To answer your question: no, currently no tests are run, since only the Docker image building c3d is built. Let me change the build definition so that `c3d/build.sh` is run instead

**Comment by tahina-pro** [comment]
Finally, I decided to move the test back to the Dockerfile, because it requires a libstdc++ more recent than what the host CI machine currently has.

**Comment by tahina-pro** [comment]
I managed to statically link with libstdc++, so the test is now back outside of the Dockerfile, at the end of `c3d/build.sh`.

**Comment by msprotz** [comment]
Thanks Tahina. Does the plugin still work with the static linking of libstc++? I'll try to fix the rest today so that we can merge.

**Comment by tahina-pro** [comment]
With or without libstdc++ statically linked, the plugin seems to fail for an independent reason.  Here's what I have:

```
: CommandLine Error: Option 'debug' registered more than once!
fatal error: error in backend: inconsistency in registered CommandLine options
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
```
I attached the whole error message.
[everparse-pull-34-plugin-error.txt](https://github.com/project-everest/everparse/files/5318582/everparse-pull-34-plugin-error.txt)

**Review by mtzguido — APPROVED** [review]
Looks good. Just one thing: how should one build C3d as a plugin now? `c3d/src/CMakeLists.txt` was changed and now doesn't include (for instance) the `include(AddLLVM)` command.

**Comment by msprotz** [comment]
My understanding of cmake is that since the top-level cmakelists includes those files, they become available in sub-cmakelists.

Note that the plugin is still built.

```
jonathan@absinthe:~/Code/quackyducky/c3d/out (protz_standalone) $ ls src/C3d.dylib 
src/C3d.dylib
jonathan@absinthe:~/Code/quackyducky/c3d/out (protz_standalone) $ ls driver/clang-c3d 
driver/clang-c3d
```

File: c3d/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+# TODO: couldn't find how to reuse the flags as picked by the build of LLVM
+# itself
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -std=c++14 -fno-rtti -fno-omit-frame-pointer")
+set(CMAKE_EXE_LINKER_FLAGS "-static-libstdc++")
+
+# Reusing the LLVM CMake stuff
+# ----------------------------
+#
+# Note: we expect users to invoke cmake -DLLVM_BUILD_DIR=path/to/llvm-project/out
```
[the remark this answers is not in the document]
Comment (msprotz) [reply c3d/CMakeLists.txt:20]: good catch thanks!

**Comment by mtzguido** [comment]
Right.. I was looking at the sub-makefile I had in `c3d/src/`. It got broken but I think it's fixed by just changing `..` to `../..`, or we could also stop supporting it. Let's merge this?

**Comment by msprotz** [comment]
Ah I missed that Makefile. Feel free to push a change to this branch to move it up one directory and fix the path, if you don't mind?

**Comment by mtzguido** [comment]
Done.

**Pull request merged**